tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,707 @@
+    }
+
+    Operand<TInt64> iLabels = cast(tf,  labels, TInt64.DTYPE);
+
+    // Try to adjust the shape so that rank of labels = rank of logits - 1.
+    Shape labelsShape = labels.asOutput().shape();
+    int labelsRank = labelsShape.numDimensions();
+
+    boolean updateShape = labelsRank != predictionsRank - 1;
+    if (updateShape) { // TODO check to see if this is right
+      Shape newShape = labelsShape.take(labelsRank - 1);
+      iLabels = tf.reshape(iLabels, tf.constant(newShape)); // flatten one dimension
+      predictions =
+          tf.reshape(
+              predictions,
+              tf.constant(
+                  new long[] {-1L, predictionsShape.size(predictionsShape.numDimensions() - 1)}));
+    }
+
+    @SuppressWarnings("unchecked")
+    Operand<T> loss = tf.nn.sparseSoftmaxCrossEntropyWithLogits(iLabels, predictions);
+    if (updateShape && predictionsRank >= 3) {
+      Shape newShape = predictionsShape.take(predictionsShape.numDimensions() - 1);
+      loss = tf.reshape(loss, tf.constant(newShape));
+    }
+    return loss;
+  }
+
+  /**
+   * Computes the squared hinge loss between labels and predictions.
+   *
+   * <p><code>loss = reduceMean(square(maximum(1 - labels * predictions, 0)))</code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets, values are expected to be -1 or 1. If binary (0 or 1) labels are *
+   *     provided, they will be converted to -1 or 1.
+   * @param predictions the predictions
+   * @param <T> the data type of the predictions and labels
+   * @return the squared hinge loss
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> squaredHinge(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = cast(tf,  labels, dataType);
+    LossTuple<T> lossTuple = LossesHelper.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = lossTuple.getTarget();
+    tLabels = lossTuple.getLabels();
+    Operand<T> one = cast(tf,  tf.constant(1), dataType);
+    Operand<T> zero = cast(tf,  tf.constant(0), dataType);
+
+    tLabels = maybeConvertLabels(tf, tLabels);
+    return tf.math.mean(
+        tf.math.square(tf.math.maximum(tf.math.sub(one, tf.math.mul(tLabels, predictions)), zero)),
+        tf.constant(-1));
+  }
+
+
+  /**
+   * Smooths binary labels
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets
+   * @param labelSmoothing A number in the range [0, 1]. When 0, no smoothing occurs. When &gt; 0,
+   *     compute the loss between the predicted labels and a smoothed version of the true labels,
+   *     where the smoothing squeezes the labels towards 0.5. Larger values of labelSmoothing
+   *     correspond to heavier smoothing.
+   * @param <T> the data type of the labels
+   * @return the smoothed binary labels
+   */
+  private static <T extends TNumber> Operand<T> smoothLabelsBinaryX(
```

Comment:
I think this would be better called `smoothBinaryLabels` as it's not specific to the binary cross entropy as far as I can tell. But it's a private method so it's not too much of an issue.

Reply:
OK